File: accsaber-business/src/main/java/de/ixsen/accsaber/business/RankedMapService.java
package de.ixsen.accsaber.business;

import de.ixsen.accsaber.business.exceptions.AccsaberOperationException;
import de.ixsen.accsaber.business.exceptions.ExceptionType;
import de.ixsen.accsaber.database.model.Category;
import de.ixsen.accsaber.database.model.maps.BeatMap;
import de.ixsen.accsaber.database.model.maps.Song;
import de.ixsen.accsaber.database.model.players.ScoreData;
import de.ixsen.accsaber.database.repositories.model.BeatMapRepository;
import de.ixsen.accsaber.database.repositories.model.CategoryRepository;
import de.ixsen.accsaber.database.repositories.model.ScoreDataRepository;
import de.ixsen.accsaber.integration.connector.BeatSaverConnector;
import de.ixsen.accsaber.integration.model.beatsaver.BeatSaverMapDifficulty;
import de.ixsen.accsaber.integration.model.beatsaver.BeatSaverSongInfo;
import de.ixsen.accsaber.integration.model.beatsaver.BeatSaverVersion;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Locale;
import java.util.Optional;

@Service
public class RankedMapService {

    private final BeatSaverConnector beatSaverConnector;
    private final BeatMapRepository beatMapRepository;
    private final ScoreDataRepository scoreDataRepository;

    private final SongService songService;
    private final CategoryRepository categoryRepository;

    @Autowired
    public RankedMapService(BeatMapRepository beatMapRepository,
                            ScoreDataRepository scoreDataRepository,
                            CategoryRepository categoryRepository,
                            BeatSaverConnector beatSaverConnector,
                            SongService songService) {
        this.beatMapRepository = beatMapRepository;
        this.scoreDataRepository = scoreDataRepository;
        this.categoryRepository = categoryRepository;
        this.beatSaverConnector = beatSaverConnector;
        this.songService = songService;
    }

    public List<BeatMap> getRankedMaps() {
        return this.beatMapRepository.findAll();
    }

    public BeatMap getRankedMap(Long leaderboardId) {
        Optional<BeatMap> optionalRankedMap = this.beatMapRepository.findById(leaderboardId);
        if (optionalRankedMap.isEmpty()) {
            throw new AccsaberOperationException(ExceptionType.RANKED_MAP_NOT_FOUND, "The ranked map with the id could not be found.");
        }
        return optionalRankedMap.get();
    }

    public void addNewRankedMapByKey(String mapKey, String difficulty, Double complexity, String categoryName) {
        BeatSaverSongInfo beatSaverSongInfo = this.beatSaverConnector.getMapInfoByKey(mapKey);
        String songHash = beatSaverSongInfo.getVersions().stream().findFirst().orElseThrow().getHash();
        this.addNewRankedMap(beatSaverSongInfo, songHash, difficulty, complexity, categoryName);
    }

    public void addNewRankedMapByHash(String songHash, String difficulty, Double complexity, String categoryName) {
        BeatSaverSongInfo beatSaverSongInfo = this.beatSaverConnector.getMapInfoByHash(songHash);
        this.addNewRankedMap(beatSaverSongInfo, songHash, difficulty, complexity, categoryName);
    }

    private void addNewRankedMap(BeatSaverSongInfo beatSaverSongInfo, String songHash, String difficulty, Double complexity, String categoryName) {
        Optional<Category> optionalCategory = this.categoryRepository.findByCategoryName(categoryName);
        if (optionalCategory.isEmpty()) {
            throw new AccsaberOperationException(ExceptionType.CATEGORY_NOT_FOUND, String.format("The category [%s] was not found.", categoryName));
        }
        Category category = optionalCategory.get();
        long leaderBoardId = this.beatSaverConnector.getScoreSaberId(songHash, this.mapDiffToId(difficulty));
        if (this.beatMapRepository.existsById(leaderBoardId)) {
            throw new AccsaberOperationException(ExceptionType.RANKED_MAP_ALREADY_EXISTS, "The ranked map with the leaderboardId " + leaderBoardId + " already exists");
        }

        BeatSaverVersion beatSaverVersion = beatSaverSongInfo.getVersions().stream()
                .filter(version -> songHash.equalsIgnoreCase(version.getHash()))
                .findAny()
                .orElseThrow(() -> new AccsaberOperationException(ExceptionType.RANKED_MAP_NOT_FOUND, String.format("BeatSaver did not return requested map with hash %s", songHash)));

        Song song = this.songService.getOrCreateSong(beatSaverSongInfo, beatSaverVersion);

        BeatSaverMapDifficulty beatSaverMapDifficulty = beatSaverVersion.getDiffs().stream()
                .filter(diff -> difficulty.equalsIgnoreCase(diff.getDifficulty()) && "Standard".equalsIgnoreCase(diff.getCharacteristic()))
                .findAny().orElseThrow(() -> new AccsaberOperationException(ExceptionType.RANKED_MAP_NOT_FOUND, String.format("Map with hash %s does not have [%s] difficulty as Standard characteristic", songHash, difficulty)));

        BeatMap beatMap = new BeatMap();
        beatMap.setLeaderboardId(leaderBoardId);
        beatMap.setMaxScore(this.calculateMaxScore(beatSaverMapDifficulty.getNotes()));
        beatMap.setSong(song);
        beatMap.setComplexity(complexity);
        beatMap.setDifficulty(difficulty);
        beatMap.setCategory(category);

        song.getBeatMaps().add(beatMap);

        song = this.songService.saveSong(song);

        this.scoreDataRepository.rankScores(leaderBoardId, beatMap.getMaxScore(), complexity);
    }

    private int mapDiffToId(String difficulty) {
        switch (difficulty.toLowerCase(Locale.ROOT)) {
            case "easy":
                return 1;
            case "normal":
                return 3;
            case "hard":
                return 5;
            case "expert":
                return 7;
            case "expertplus":
                return 9;
        }
        return -1;
    }

    public void removeRankedMap(Long leaderboardId) {
        Optional<BeatMap> map = this.beatMapRepository.findById(leaderboardId);
        if (map.isEmpty()) {
            throw new AccsaberOperationException(ExceptionType.RANKED_MAP_NOT_FOUND,
                    "The ranked map [" + leaderboardId + "] does not currently exist.");
        }

        BeatMap beatMap = map.get();
        Song song = beatMap.getSong();
        song.getBeatMaps().remove(beatMap);
        beatMap.setSong(null);
        if (song.getBeatMaps().size() == 0) {
            this.songService.removeSong(song.getSongHash());
        } else {
            this.songService.saveSong(song);
        }

        List<ScoreData> unrankedScores = this.scoreDataRepository.findAllByLeaderboardId(leaderboardId);
        unrankedScores.forEach(score -> score.setRankedScore(false));
        this.scoreDataRepository.saveAll(unrankedScores);
    }


    // There has to be a better way to do this
    private int calculateMaxScore(int noteCount) {
        int score = 0;
        for (int i = 1; i <= noteCount; i++) {

            if (i <= 1) {
                score += 1;
            } else if (i <= 5) {
                score += 2;
            } else if (i <= 13) {
                score += 4;
            } else {
                break;
            }
        }
        score *= 115;
        if (noteCount > 13) {
            score += (8 * (noteCount - 13)) * 115;
        }

        return score;
    }


}

File: accsaber-business/src/main/kotlin/de/ixsen/accsaber/business/RankedMapService.kt
package de.ixsen.accsaber.business

import de.ixsen.accsaber.business.exceptions.AccsaberOperationException
import de.ixsen.accsaber.business.exceptions.ExceptionType
import de.ixsen.accsaber.database.model.Category
import de.ixsen.accsaber.database.model.maps.BeatMap
import de.ixsen.accsaber.database.model.players.ScoreData
import de.ixsen.accsaber.database.repositories.model.BeatMapRepository
import de.ixsen.accsaber.database.repositories.model.CategoryRepository
import de.ixsen.accsaber.database.repositories.model.ScoreDataRepository
import de.ixsen.accsaber.integration.connector.BeatSaverConnector
import de.ixsen.accsaber.integration.model.beatsaver.BeatSaverMapDifficulty
import de.ixsen.accsaber.integration.model.beatsaver.BeatSaverSongInfo
import de.ixsen.accsaber.integration.model.beatsaver.BeatSaverVersion
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.function.Consumer

@Service
class RankedMapService @Autowired constructor(
    private val beatMapRepository: BeatMapRepository,
    private val scoreDataRepository: ScoreDataRepository,
    private val categoryRepository: CategoryRepository,
    private val beatSaverConnector: BeatSaverConnector,
    private val songService: SongService
) {
    fun getAllRankedMaps(): List<BeatMap> {
        return beatMapRepository.findAll()
    }

    fun getAllRankedMapsForCategory(category: Category): List<BeatMap> {
        return this.beatMapRepository.findAllForCategory(category)
    }

    fun getRankedMap(leaderboardId: Long): BeatMap {
        val optionalRankedMap = beatMapRepository.findById(leaderboardId)
        if (optionalRankedMap.isEmpty) {
            throw AccsaberOperationException(ExceptionType.RANKED_MAP_NOT_FOUND, "The ranked map with the id could not be found.")
        }
        return optionalRankedMap.get()
    }

    fun addNewRankedMapByKey(mapKey: String, difficulty: String, complexity: Double, categoryName: String) {
        val beatSaverSongInfo = beatSaverConnector.getMapInfoByKey(mapKey)!!
        val songHash = beatSaverSongInfo.versions.stream().findFirst().orElseThrow().hash
        addNewRankedMap(beatSaverSongInfo, songHash!!, difficulty, complexity, categoryName)
    }

    fun addNewRankedMapByHash(songHash: String, difficulty: String, complexity: Double, categoryName: String) {
        val beatSaverSongInfo = beatSaverConnector.getMapInfoByHash(songHash)!!
        addNewRankedMap(beatSaverSongInfo, songHash, difficulty, complexity, categoryName)
    }

    private fun addNewRankedMap(beatSaverSongInfo: BeatSaverSongInfo, songHash: String, difficulty: String, complexity: Double, categoryName: String) {
        val optionalCategory = categoryRepository.findByCategoryName(categoryName)
        if (optionalCategory.isEmpty) {
            throw AccsaberOperationException(ExceptionType.CATEGORY_NOT_FOUND, String.format("The category [%s] was not found.", categoryName))
        }
        val category = optionalCategory.get()
        val leaderBoardId = beatSaverConnector.getScoreSaberId(songHash, mapDiffToId(difficulty))
        if (beatMapRepository.existsById(leaderBoardId)) {
            throw AccsaberOperationException(ExceptionType.RANKED_MAP_ALREADY_EXISTS, "The ranked map with the leaderboardId $leaderBoardId already exists")
        }
        val beatSaverVersion = beatSaverSongInfo.versions.stream()
            .filter { version: BeatSaverVersion -> songHash.equals(version.hash, ignoreCase = true) }
            .findAny()
            .orElseThrow {
                AccsaberOperationException(
                    ExceptionType.RANKED_MAP_NOT_FOUND,
                    String.format("BeatSaver did not return requested map with hash %s", songHash)
                )
            }
        var song = songService.getOrCreateSong(beatSaverSongInfo, beatSaverVersion)
        val beatSaverMapDifficulty = beatSaverVersion.diffs.stream()
            .filter { diff: BeatSaverMapDifficulty -> difficulty.equals(diff.difficulty, ignoreCase = true) && "Standard".equals(diff.characteristic, ignoreCase = true) }
            .findAny().orElseThrow {
                AccsaberOperationException(
                    ExceptionType.RANKED_MAP_NOT_FOUND,
                    String.format("Map with hash %s does not have [%s] difficulty as Standard characteristic", songHash, difficulty)
                )
            }
        val beatMap = BeatMap(
            leaderBoardId,
            calculateMaxScore(beatSaverMapDifficulty.notes!!),
            song,
            difficulty,
            complexity,
            category
        )

        song.beatMaps.add(beatMap)
        song = songService.saveSong(song)
        scoreDataRepository.rankScores(leaderBoardId, beatMap.maxScore, complexity)
    }

    private fun mapDiffToId(difficulty: String): Int {
        when (difficulty.lowercase()) {
            "easy" -> return 1
            "normal" -> return 3
            "hard" -> return 5
            "expert" -> return 7
            "expertplus" -> return 9
        }
        return -1
    }

    fun removeRankedMap(leaderboardId: Long) {
        val map = beatMapRepository.findById(leaderboardId)
        if (map.isEmpty) {
            throw AccsaberOperationException(
                ExceptionType.RANKED_MAP_NOT_FOUND,
                "The ranked map [$leaderboardId] does not currently exist."
            )
        }
        val beatMap = map.get()
        val song = beatMap.song
        song.beatMaps.remove(beatMap)
        if (song.beatMaps.size == 0) {
            songService.removeSong(song.songHash)
        } else {
            songService.saveSong(song)
        }
        val unrankedScores = scoreDataRepository.findAllByLeaderboardId(leaderboardId)
        unrankedScores.forEach(Consumer { score: ScoreData -> score.isRankedScore = false })
        scoreDataRepository.saveAll(unrankedScores)
    }

    // There has to be a better way to do this
    private fun calculateMaxScore(noteCount: Int): Int {
        var score = 0L
        for (i in 1..noteCount) {
            score += if (i <= 1) {
                1
            } else if (i <= 5) {
                2
            } else if (i <= 13) {
                4
            } else {
                break
            }
        }
        score *= 115
        if (noteCount > 13) {
            score += 8 * (noteCount - 13) * 115
        }
        return score.toInt()
    }

    fun updateComplexity(leaderboardId: String, complexity: Double) {
        val beatMap = this.beatMapRepository.findById(leaderboardId.toLong())
            .orElseThrow { AccsaberOperationException(ExceptionType.RANKED_MAP_NOT_FOUND, "The ranked map [$leaderboardId] does not currently exist.") }

        beatMap.complexity = complexity
        this.beatMapRepository.save(beatMap)
    }
}